File: frontend/src/security/securityHeaders.ts
/**
 * Security headers configuration and middleware
 * Implements comprehensive security headers for OWASP compliance
 */

/**
 * Content Security Policy configuration
 */
export interface CSPConfig {
  defaultSrc?: string[];
  scriptSrc?: string[];
  styleSrc?: string[];
  imgSrc?: string[];
  fontSrc?: string[];
  connectSrc?: string[];
  mediaSrc?: string[];
  objectSrc?: string[];
  frameSrc?: string[];
  frameAncestors?: string[];
  formAction?: string[];
  baseUri?: string[];
  reportUri?: string;
  upgradeInsecureRequests?: boolean;
}

/**
 * Generate Content Security Policy string
 */
export function generateCSP(config: CSPConfig): string {
  const directives: string[] = [];

  // Helper to format directive values
  const formatDirective = (name: string, values?: string[]): void => {
    if (values && values.length > 0) {
      directives.push(`${name} ${values.join(' ')}`);
    }
  };

  formatDirective('default-src', config.defaultSrc || ["'self'"]);
  formatDirective('script-src', config.scriptSrc);
  formatDirective('style-src', config.styleSrc);
  formatDirective('img-src', config.imgSrc);
  formatDirective('font-src', config.fontSrc);
  formatDirective('connect-src', config.connectSrc);
  formatDirective('media-src', config.mediaSrc);
  formatDirective('object-src', config.objectSrc || ["'none'"]);
  formatDirective('frame-src', config.frameSrc);
  formatDirective('frame-ancestors', config.frameAncestors || ["'none'"]);
  formatDirective('form-action', config.formAction || ["'self'"]);
  formatDirective('base-uri', config.baseUri || ["'self'"]);

  if (config.reportUri) {
    directives.push(`report-uri ${config.reportUri}`);
  }

  if (config.upgradeInsecureRequests) {
    directives.push('upgrade-insecure-requests');
  }

  return directives.join('; ');
}

/**
 * Get API URL for CSP configuration
 */
function getApiUrlForCSP(): string {
  const apiUrl = import.meta.env.VITE_API_URL || (
    import.meta.env.DEV 
      ? 'http://localhost:3000'
      : 'https://api.example.com'
  );
  
  try {
    const url = new URL(apiUrl);
    return `${url.protocol}//${url.host}`;
  } catch {
    return "'self'";
  }
}

/**
 * Default CSP configuration for production
 */
export const defaultCSPConfig: CSPConfig = {
  defaultSrc: ["'self'"],
  scriptSrc: ["'self'", "'strict-dynamic'"],
  styleSrc: ["'self'", "'unsafe-inline'"], // Consider using nonces instead
  imgSrc: ["'self'", 'data:', 'https:'],
  fontSrc: ["'self'"],
  connectSrc: ["'self'", getApiUrlForCSP()],
  mediaSrc: ["'none'"],
  objectSrc: ["'none'"],
  frameAncestors: ["'none'"],
  formAction: ["'self'"],
  baseUri: ["'self'"],
  upgradeInsecureRequests: true,
};

/**
 * Strict CSP configuration (recommended)
 */
export const strictCSPConfig: CSPConfig = {
  defaultSrc: ["'none'"],
  scriptSrc: ["'self'", "'strict-dynamic'"],
  styleSrc: ["'self'"],
  imgSrc: ["'self'"],
  fontSrc: ["'self'"],
  connectSrc: ["'self'"],
  mediaSrc: ["'none'"],
  objectSrc: ["'none'"],
  frameAncestors: ["'none'"],
  formAction: ["'self'"],
  baseUri: ["'self'"],
  upgradeInsecureRequests: true,
};

/**
 * Security headers configuration
 */
export interface SecurityHeaders {
  'Content-Security-Policy'?: string;
  'X-Content-Type-Options'?: string;
  'X-Frame-Options'?: string;
  'X-XSS-Protection'?: string;
  'Strict-Transport-Security'?: string;
  'Referrer-Policy'?: string;
  'Permissions-Policy'?: string;
  'Cross-Origin-Embedder-Policy'?: string;
  'Cross-Origin-Opener-Policy'?: string;
  'Cross-Origin-Resource-Policy'?: string;
}

/**
 * Get recommended security headers
 */
export function getSecurityHeaders(options: {
  cspConfig?: CSPConfig;
  enableHSTS?: boolean;
  hstsMaxAge?: number;
  enableFrameOptions?: boolean;
  frameOptions?: 'DENY' | 'SAMEORIGIN';
  referrerPolicy?: string;
} = {}): SecurityHeaders {
  const {
    cspConfig = defaultCSPConfig,
    enableHSTS = true,
    hstsMaxAge = 31536000, // 1 year
    enableFrameOptions = true,
    frameOptions = 'DENY',
    referrerPolicy = 'strict-origin-when-cross-origin',
  } = options;

  const headers: SecurityHeaders = {
    'Content-Security-Policy': generateCSP(cspConfig),
    'X-Content-Type-Options': 'nosniff',
    'X-XSS-Protection': '1; mode=block',
    'Referrer-Policy': referrerPolicy,
    'Permissions-Policy': 'geolocation=(), microphone=(), camera=(), payment=()',
    'Cross-Origin-Embedder-Policy': 'require-corp',
    'Cross-Origin-Opener-Policy': 'same-origin',
    'Cross-Origin-Resource-Policy': 'same-origin',
  };

  if (enableHSTS) {
    headers['Strict-Transport-Security'] = `max-age=${hstsMaxAge}; includeSubDomains; preload`;
  }

  if (enableFrameOptions) {
    headers['X-Frame-Options'] = frameOptions;
  }

  return headers;
}

/**
 * Add security headers to HTML meta tags
 */
export function addSecurityMetaTags(): void {
  const head = document.head;

  // CSP meta tag
  const cspMeta = document.createElement('meta');
  cspMeta.httpEquiv = 'Content-Security-Policy';
  cspMeta.content = generateCSP(defaultCSPConfig);
  head.appendChild(cspMeta);

  // Other security-related meta tags
  const metaTags = [
    { httpEquiv: 'X-Content-Type-Options', content: 'nosniff' },
    { httpEquiv: 'X-Frame-Options', content: 'DENY' },
    { httpEquiv: 'X-XSS-Protection', content: '1; mode=block' },
    { name: 'referrer', content: 'strict-origin-when-cross-origin' },
  ];

  metaTags.forEach(tag => {
    const meta = document.createElement('meta');
    if (tag.httpEquiv) {
      meta.httpEquiv = tag.httpEquiv;
    } else {
      meta.name = tag.name!;
    }
    meta.content = tag.content;
    head.appendChild(meta);
  });
}

/**
 * Vite plugin for security headers
 */
export const securityHeadersPlugin = () => ({
  name: 'security-headers',
  configureServer(server: any) {
    server.middlewares.use((req: any, res: any, next: any) => {
      const headers = getSecurityHeaders();
      
      Object.entries(headers).forEach(([key, value]) => {
        if (value) {
          res.setHeader(key, value);
        }
      });

      next();
    });
  },
  transformIndexHtml(html: string) {
    // Add security meta tags to HTML
    const securityMeta = `
    <meta http-equiv="Content-Security-Policy" content="${generateCSP(defaultCSPConfig)}">
    <meta http-equiv="X-Content-Type-Options" content="nosniff">
    <meta http-equiv="X-Frame-Options" content="DENY">
    <meta http-equiv="X-XSS-Protection" content="1; mode=block">
    <meta name="referrer" content="strict-origin-when-cross-origin">
    `.trim();

    return html.replace('</head>', `${securityMeta}\n  </head>`);
  },
});

/**
 * Generate nonce for inline scripts/styles
 */
export function generateNonce(): string {
  const array = new Uint8Array(16);
  crypto.getRandomValues(array);
  return btoa(String.fromCharCode.apply(null, Array.from(array)));
}

/**
 * CSP violation reporter
 */
export class CSPViolationReporter {
  private violations: any[] = [];
  private reportEndpoint: string;

  constructor(reportEndpoint: string) {
    this.reportEndpoint = reportEndpoint;
    this.setupListener();
  }

  private setupListener(): void {
    document.addEventListener('securitypolicyviolation', (e) => {
      const violation = {
        timestamp: new Date().toISOString(),
        documentUri: e.documentURI,
        blockedUri: e.blockedURI,
        violatedDirective: e.violatedDirective,
        effectiveDirective: e.effectiveDirective,
        originalPolicy: e.originalPolicy,
        sourceFile: e.sourceFile,
        lineNumber: e.lineNumber,
        columnNumber: e.columnNumber,
        statusCode: e.statusCode,
        userAgent: navigator.userAgent,
      };

      this.violations.push(violation);
      this.reportViolation(violation);
    });
  }

  private async reportViolation(violation: any): Promise<void> {
    try {
      await fetch(this.reportEndpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(violation),
      });
    } catch (error) {
      console.error('Failed to report CSP violation:', error);
    }
  }

  getViolations(): any[] {
    return [...this.violations];
  }

  clearViolations(): void {
    this.violations = [];
  }
}

/**
 * Security headers validator for testing
 */
export function validateHeaders(headers: Headers): {
  valid: boolean;
  missing: string[];
  warnings: string[];
} {
  const required = [
    'Content-Security-Policy',
    'X-Content-Type-Options',
    'X-Frame-Options',
    'Strict-Transport-Security',
  ];

  const missing = required.filter(header => !headers.has(header));
  const warnings: string[] = [];

  // Validate CSP
  const csp = headers.get('Content-Security-Policy');
  if (csp) {
    if (csp.includes('unsafe-inline') && !csp.includes('nonce-')) {
      warnings.push('CSP uses unsafe-inline without nonce');
    }
    if (csp.includes('unsafe-eval')) {
      warnings.push('CSP uses unsafe-eval');
    }
    if (!csp.includes('upgrade-insecure-requests')) {
      warnings.push('CSP missing upgrade-insecure-requests');
    }
  }

  // Validate HSTS
  const hsts = headers.get('Strict-Transport-Security');
  if (hsts && !hsts.includes('includeSubDomains')) {
    warnings.push('HSTS missing includeSubDomains');
  }

  return {
    valid: missing.length === 0,
    missing,
    warnings,
  };
}

/**
 * HTTPS enforcement utilities
 */
export const httpsEnforcement = {
  /**
   * Redirect to HTTPS if not already
   */
  enforceHTTPS(): void {
    if (window.location.protocol === 'http:' && window.location.hostname !== 'localhost') {
      window.location.href = window.location.href.replace('http:', 'https:');
    }
  },

  /**
   * Check if current connection is secure
   */
  isSecure(): boolean {
    return window.location.protocol === 'https:' || window.location.hostname === 'localhost';
  },

  /**
   * Validate URL uses HTTPS
   */
  validateHTTPS(url: string): boolean {
    try {
      const parsed = new URL(url);
      return parsed.protocol === 'https:' || parsed.hostname === 'localhost';
    } catch {
      return false;
    }
  },
};

File: backend/services/shared/middleware/security_headers.py
"""
Security Headers Middleware
Implements comprehensive security headers including CSP for backend services
"""

from typing import Dict, List, Optional, Callable
from fastapi import Request, Response
from fastapi.responses import JSONResponse
import secrets
import hashlib
import base64
import json
from datetime import datetime
import logging

logger = logging.getLogger(__name__)


class SecurityHeadersMiddleware:
    """
    Middleware to add security headers to all responses
    """
    
    def __init__(
        self,
        app,
        csp_config: Optional[Dict[str, List[str]]] = None,
        enable_nonce: bool = True,
        report_uri: Optional[str] = None,
        report_only: bool = False
    ):
        self.app = app
        self.csp_config = csp_config or self._get_default_csp()
        self.enable_nonce = enable_nonce
        self.report_uri = report_uri
        self.report_only = report_only
    
    async def __call__(self, request: Request, call_next: Callable) -> Response:
        # Generate nonce for this request if enabled
        nonce = None
        if self.enable_nonce:
            nonce = self._generate_nonce()
            request.state.csp_nonce = nonce
        
        # Process the request
        response = await call_next(request)
        
        # Add security headers
        self._add_security_headers(response, nonce)
        
        return response
    
    def _get_default_csp(self) -> Dict[str, List[str]]:
        """Get default CSP configuration"""
        return {
            "default-src": ["'self'"],
            "script-src": ["'self'", "'strict-dynamic'"],
            "style-src": ["'self'", "'unsafe-inline'"],  # Required for some frameworks
            "img-src": ["'self'", "data:", "https:"],
            "font-src": ["'self'"],
            "connect-src": ["'self'"],
            "media-src": ["'self'"],
            "object-src": ["'none'"],
            "frame-src": ["'none'"],
            "frame-ancestors": ["'none'"],
            "base-uri": ["'self'"],
            "form-action": ["'self'"],
            "upgrade-insecure-requests": []
        }
    
    def _generate_nonce(self) -> str:
        """Generate a cryptographically secure nonce"""
        return base64.b64encode(secrets.token_bytes(16)).decode('utf-8')
    
    def _build_csp_header(self, nonce: Optional[str] = None) -> str:
        """Build the CSP header value"""
        directives = []
        
        for directive, sources in self.csp_config.items():
            if directive == "script-src" and nonce:
                # Add nonce to script-src
                sources_with_nonce = sources + [f"'nonce-{nonce}'"]
                directives.append(f"{directive} {' '.join(sources_with_nonce)}")
            else:
                if sources:
                    directives.append(f"{directive} {' '.join(sources)}")
                else:
                    directives.append(directive)
        
        csp = "; ".join(directives)
        
        if self.report_uri:
            csp += f"; report-uri {self.report_uri}"
        
        return csp
    
    def _add_security_headers(self, response: Response, nonce: Optional[str] = None):
        """Add all security headers to the response"""
        # Content Security Policy
        csp_header_name = "Content-Security-Policy-Report-Only" if self.report_only else "Content-Security-Policy"
        response.headers[csp_header_name] = self._build_csp_header(nonce)
        
        # Other security headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        response.headers["Permissions-Policy"] = (
            "camera=(), microphone=(), geolocation=(), payment=()"
        )
        
        # HSTS (only in production)
        if not self._is_development():
            response.headers["Strict-Transport-Security"] = (
                "max-age=31536000; includeSubDomains; preload"
            )


    def _is_development(self) -> bool:
        """Check if running in development mode"""
        import os
        return os.getenv("ENVIRONMENT", "development") == "development"


class CSPReportHandler:
    """
    Handler for CSP violation reports
    """
    
    def __init__(self, storage_backend: Optional[Callable] = None):
        self.storage_backend = storage_backend or self._default_storage
        self.logger = logging.getLogger(__name__)
    
    async def handle_report(self, request: Request) -> JSONResponse:
        """Handle CSP violation report"""
        try:
            # Parse the report
            body = await request.body()
            report_data = json.loads(body)
            
            # Extract CSP report
            csp_report = report_data.get("csp-report", {})
            
            # Log the violation
            self.logger.warning(
                "CSP Violation",
                extra={
                    "document_uri": csp_report.get("document-uri"),
                    "violated_directive": csp_report.get("violated-directive"),
                    "blocked_uri": csp_report.get("blocked-uri"),
                    "source_file": csp_report.get("source-file"),
                    "line_number": csp_report.get("line-number"),
                    "column_number": csp_report.get("column-number"),
                    "timestamp": datetime.utcnow().isoformat()
                }
            )
            
            # Store the report
            await self.storage_backend(csp_report)
            
            return JSONResponse(
                status_code=204,
                content=None
            )
            
        except Exception as e:
            self.logger.error(f"Failed to process CSP report: {str(e)}")
            return JSONResponse(
                status_code=500,
                content={"error": "Failed to process report"}
            )
    
    async def _default_storage(self, report: dict):
        """Default storage implementation (logs only)"""
        self.logger.info(f"CSP Report: {json.dumps(report, indent=2)}")


def get_production_csp_config() -> Dict[str, List[str]]:
    """Get production CSP configuration"""
    return {
        "default-src": ["'self'"],
        "script-src": ["'self'"],
        "style-src": ["'self'"],
        "img-src": ["'self'", "https:"],
        "font-src": ["'self'"],
        "connect-src": ["'self'", "https://api.yourcompany.com"],
        "media-src": ["'self'"],
        "object-src": ["'none'"],
        "frame-src": ["'none'"],
        "frame-ancestors": ["'none'"],
        "base-uri": ["'self'"],
        "form-action": ["'self'"],
        "upgrade-insecure-requests": [],
        "block-all-mixed-content": []
    }


def get_development_csp_config() -> Dict[str, List[str]]:
    """Get development CSP configuration (more permissive)"""
    return {
        "default-src": ["'self'"],
        "script-src": ["'self'", "'unsafe-inline'", "'unsafe-eval'", "http://localhost:*"],
        "style-src": ["'self'", "'unsafe-inline'"],
        "img-src": ["'self'", "data:", "https:", "http:"],
        "font-src": ["'self'", "data:"],
        "connect-src": ["'self'", "http://localhost:*", "ws://localhost:*"],
        "media-src": ["'self'"],
        "object-src": ["'none'"],
        "frame-src": ["'self'"],
        "frame-ancestors": ["'self'"],
        "base-uri": ["'self'"],
        "form-action": ["'self'"]
    }


# FastAPI app setup example:
"""
from fastapi import FastAPI
from .security_headers import SecurityHeadersMiddleware, CSPReportHandler, get_production_csp_config

app = FastAPI()

# Add security headers middleware
app.add_middleware(
    SecurityHeadersMiddleware,
    csp_config=get_production_csp_config(),
    enable_nonce=True,
    report_uri="/api/csp-report",
    report_only=False  # Set to True initially to test
)

# Add CSP report endpoint
csp_handler = CSPReportHandler()

@app.post("/api/csp-report")
async def csp_report(request: Request):
    return await csp_handler.handle_report(request)
"""